src/compiler/checker.ts

```diff
@@ -20258,7 +20264,7 @@ namespace ts {
                         getArrayifiedType(checkExpressionWithContextualType((<SpreadElement>arg).expression, restType, context, CheckMode.Normal));
                 }
             }
-            const contextualType = getIndexTypeOfType(restType, IndexKind.Number) || anyType;
+            const contextualType = getIndexedAccessType(restType, numberType);

```

Comment:
```suggestion
            const contextualType = getIndexedAccessType(restType, getLiteralType(index));
```
?

Reply:
In principle yes, but not exactly like that. Just pushed a new commit.